### ISSUE TITLE
[QUESTION] How to re-do forward pass

### ISSUE THREAD

**Issue by gahdritz** [microsoft/DeepSpeed#1600]
Is it possible during distributed DeepSpeed training to "reset" the model's gradients, activation checkpoints, etc. and repeat the forward pass from scratch in a single rank? In my use case, I have to conditionally change the model's inputs depending on their respective outputs, so, ideally, I'd be able to run the model once using one option, perform some kind of check, and then use a different input instead if necessary, all without running a backward pass.

**Comment by cli99** [comment]
Here is a typical training workflow. https://www.deepspeed.ai/getting-started/#training. Is your case like this?

    #forward() method
    loss = model_engine(input)
    next_input = some_check(loss)
    next_loss = model_engine(next_input)

    #runs backpropagation
    model_engine.backward(next_loss)

    #weight update
    model_engine.step()

Can you elaborate more on your use case where repeating the forward pass only happens on a single rank?

**Comment by gahdritz** [comment]
This is related to issue #1496.

My proposed fix to the problem there was just to return a zero tensor when the loss becomes NaN after the forward pass. However, during distributed training, this doesn't work: each of the ranks needs to update the same subset of the model's parameters, or training hangs.

My next thought, then, was to attempt a forward pass, and then, if the loss becomes NaN, "reset" the forward pass like I describe in this issue and try again with e.g. zero inputs that yield a real loss. That way, I can obtain a zero loss tensor that nevertheless interacts with the same subset of parameters.

Without the ability to "reset" the forward pass as I describe, I've had to resort to running the forward pass once with grad disabled, checking if the result is NaN, and then running the real forward pass with either the real or "NaN-friendly" inputs. This approach comes at the cost of an additional forward pass, though, so the ability to "reset" would be nice.

**Comment by yaozhewei** [comment]
There are multiple ways to realize the functionality you want to achieve the goal you want.

1. Using the way cli99 mentioned. 
If you are afraid that there are some cached activation etc, you can do something like
```
#forward() method
with torch.no_grad(): 
   # This won't cache anything
    loss = model_engine(input)
next_input = some_check(loss)
next_loss = model_engine(next_input)

#runs backpropagation
model_engine.backward(next_loss)

#weight update
model_engine.step()
```

2. You can actually  remove those NaN from loss so that they would affect backward pass.
Here we give pseudo-code and a simple example to illustrate this.
```
loss = [1, NaN, 1, Nan]
new_loss = torch.sum(loss[0, 2])
model_engine.backward(new_loss )
model_engine.step()
```

Here is an pytorch example we have tested for you
```
>>> A  = torch.nn.Linear(3, 1)
>>> input = torch.randn(2, 3)
>>> output = A(input)
>>> output
tensor([[0.3624],
        [0.8110]], grad_fn=<AddmmBackward>)
>>> loss = output[0][0] ** 2
>>> loss.backward()
>>> A.weight.grad.data
tensor([[-0.6942,  0.5204,  0.1487]])
>>> A.zero_grad()
>>> output = A(input[:1, :])
>>> loss = output ** 2
>>> loss.backward()
>>> A.weight.grad.data
tensor([[-0.6942,  0.5204,  0.1487]])
```
As you can see, the two gradient values are matched. However, one thing you need to pay attention to is that: why you are doing distributed training, the gradient will be averaged across all GPUs. When you do such a loss filtering, the effective batch size on each GPU is different (for example, after filtering, the actually batch size of GPU0--3 might be 1, 2, 3, 4). You may need to take this into consideration.

**Comment by gahdritz** [comment]
Thanks for the reply.

1. Part of the problem is that my model is extremely large and contains many activation checkpoints. Even without the stored gradients from the first pass, the stored inputs in each activation checkpoint alone would be a major memory bottleneck. It seems to me that the code snippet you suggest is my current solution, which I described in my previous comment. The problem is that, this way, you need to run an entire additional forward pass no matter what the input is. Ideally, it would only be necessary in the rare cases where the loss turns out to be NaN.
2. Unfortunately, my model is big enough to require batch sizes of just 1 on most hardware, so this wouldn't work in my case.

**Comment by loadams** [comment]
Hi @gahdritz - I'm closing this issue as it appears to be stale.  If you are still hitting this issue, please open a new one and link this and we will prioritize looking at it.
